Rework vectorization analysis for View

csarofeen opened the pull request:
> Updated version of https://github.com/csarofeen/pytorch/pull/2038 This is ready to be reviewed.

Need to update comments
Need to look into a few regressions - Vectorization factor changed in some cases, need to see which ones changed correctly/incorrectly. There's probably a bug somewhere computing the vectorization factor. Strangely these changes seem to have helped more than hurt, so will be interesting to see what's going on, maybe it's vectorizing cases where it shouldn't be correct to.

![image](https://user-images.githubusercontent.com/22205833/200325959-dc0ea643-8d4e-4f51-b8a2-907a9c85844b.png)

@naoyam @zasdfgbnm this should actually be ready for review if you want to start. I need to update the comments but all this PR does is switch to keeping full factors of partial mappings through the graph (also symbolically), and avoid propagating through non divisible splits.

The hardest part of reviewing this PR is reviewing this simple fraction system I built, all it supports is multiplication and division, and has some (simple) simplification functions. This is far from a math algebraic simplifier, but should be sufficient for our purposes for a while.

csarofeen committed:
One failing test.

csarofeen committed:
Disable conservative vectorization validation.

csarofeen commented on torch/csrc/jit/codegen/cuda/scheduler/pointwise.cpp:
```diff
@@ -243,7 +243,9 @@ std::shared_ptr<PointwiseParams> getPointwiseHeuristics(
-      for (const auto break_point_i : c10::irange(ref_root.size())) {
+      // Don't check the inner most dimension, scheduler assumes there's always
+      // an rhs
+      for (const auto break_point_i : c10::irange(ref_root.size() - 1)) {
```
Unrelated fix, but relatively small so leaving here.

csarofeen committed:
Cache vectorize mapper use in heuristics.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
Merge conflict.

naoyam commented on torch/csrc/jit/codegen/cuda/test/test_gpu2.cpp:
```diff
@@ -4191,8 +4191,8 @@ TEST_F(NVFuserTest, FusionVectorizeMisalignedPointwiseMergeSymbolicPass_CUDA) {
 
-  auto tv0 = makeSymbolicTensor(kNumDims);
-  auto tv1 = makeSymbolicTensor(kNumDims);
+  auto tv0 = makeContigTensor(kNumDims);
```
Haven't looked at the changes of this PR yet, but why were these changes made? Does this PR have anything to do with MisalignedVectorize?

naoyam replied:
Tried the tests in this file. It seems none needs to be changed except for `FusionVectorization3`.

csarofeen replied:
I don't remember, doesn't misaligned vectorization recheck the start boundary for every row independently? Maybe that was my poor assumption.

csarofeen replied:
Even if it did, if the inner most stride wasn't 1 then we couldn't use it right?

naoyam replied:
There's some runtime checks for misaligned vectorization, and I think that the innermost stride is one of them. So, it should fail at the run time if a non-stride-1 tensor was given.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,207 @@
-    Fusion* fusion,
+// Projects IterDomains through the fusion starting at provided reference. IDs
+// in the reference are expected to be "contiguous", simply means dimensions
+// that the iter domains are consecutive and next to eachother in the reference.
```
typo: each other

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,207 @@
+// Input: T0[i0, i2]
+// Reference: T5[i0, i1, i2]
+// If we want to base the vectorization size on the reference being contiguous
+// in a 1D scheduler, we'd start the proces on the reference with {i0, i1, i2}.
```
typo: process

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,207 @@
+// Refrence [i0, i1, i2]
+// Keeping the ordering relative to the reference also allows us to look though
+// transpose operations without missing out in a case like this that the
+// reference and input are consistently ordered so we can look at i0*i1*i2 for
```
Are the reference and input consistently ordered? They appear to have different orderings. The input and intermediate do look consistently ordered.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+  // T1[i0, b1, i2] = broadcast(T0)
+  // T2[i0, i1, i2]
+  // T3[i0, i1, i2] = T1 + T2
+  // and we're propogating from T3 with {i0, i1, i2}
```
typo: propagating

csarofeen replied:
Some day I'll learn to spell, thanks :-)

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        // everything to the left of inner is not contiguously merged.
+        //
+        // Clear left of inner
+        clear_left_of(find_inner_it);
```
The below cases have `continue`. Do we want one here too? Still trying to understand it's just optional.

It'd be really helpful to understand if you could add a brief comment at each path saying if it's possible to have a `continue`. The function and the outer loop of `transform_exprs` are quite large. Would be helpful to have more comments on the overall control flow.

csarofeen replied:
Will add a comment. We want to continue processing the statement here because inner maps so we might have a partial inner mapping to continue with. We just don't have outer so anything left of inner can't partially map.

csarofeen replied:
Comment added will be:
```
      // There are a few continue conditions below that stop processing the
      // expression. Continue is used when we cannot map any information even
      // partial of the expression.
      //
      // If inner is unmapped.
      // If neither inner or outer are mapped.
      // If inner and outer are mapped but inner is mapped outside of outer.
      //
      // If the above conditions aren't hit, then we want to continue
      // propagation of a partial mapping through the expressions.
      ```

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        partial_projected_extent_[split->in()] =
+            getMaybePartialMappedExtent(split->inner());
+        // Clear to the left of split in since it partially maps
+        ids.erase(ids.begin(), ids.begin() + inner_pos);
```
IIUC, this erasing is redundant as we already cleared them out at lines 200 and 205. Am I right? Just want to confirm if I understand this correctly.

csarofeen replied:
Yeah, I believe this is redundant. Will remove.

csarofeen replied:
Backwards split:
First Check: If inner was marked as partially mapped remove to the left of inner
Second Check: If in is partially mapped remove to the left of inner

Backwards merge:
If out is partially mapped, remove everything left of out, mark partially mapped

So I do think we can remove the first check in backwards split. I'm going to comment it's likely redundant, but I'm scared to modify this code too much 🤣 so I'd rather leave some redundant code for now but mark it as such.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+            // Partial mapping on out is bigger than inner, so extend partial
+            // mapping to outer and mark inner as fully mapped.
+            //
+            // TODO: Should this be ceilDiv? Normal div is more conservative, so
```
I don't think it should be ceilDiv. It would incorrectly enlarge the contig extent.

zasdfgbnm replied:
I think it should neither be ceilDiv nor div. It should be something like
```C++
if (divisible) {
  take the smaller
} else {
  erase both inner and outer
}
```
See also https://github.com/csarofeen/pytorch/pull/2124#discussion_r1004991114

csarofeen replied:
I think div is correct here. We're trying to get a total inner most mapped size to help make a decision. That inner most dimension size must be a multiple, and this partially mapped portion can be part of that decision. However, I'm open minded I'm getting this wrong.

csarofeen replied:
Going to think about this more.

naoyam reviewed (commented):
Some more comments

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,207 @@
+// Keeping the ordering relative to the reference also allows us to look though
+// transpose operations without missing out in a case like this that the
+// reference and input are consistently ordered so we can look at i0*i1*i2 for
+// our vector multiple even though there are transposes in between them.
```
I think you meant the following?
```C++
// Input:[i0, i1, i2]
// Intermediate: [i1, i0, i2]
// Refrence [i0, i1, i2]
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        find_outer_it = std::find(ids.begin(), ids.end(), split->outer());
+        outer_pos = find_outer_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_outer_it);
```
nit:
```suggestion
        outer_pos = std::distance(ids.begin(), find_outer_it);
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        find_inner_it = std::find(ids.begin(), ids.end(), split->inner());
+        inner_pos = find_inner_it == ids.end()
+            ? ids.size()
+            : std::distance(ids.begin(), find_inner_it);
```
nit:
```suggestion
        inner_pos = std::distance(ids.begin(), find_inner_it);
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+          // of a continuous mapping. There are cases where we could have
+          // reversible operations e.g.:
+          //    [id{3} id{5} id{6}] -> merge(1, 0)
+          // -> [id{5*3} id{6}] -> split(0, 5)
```
I think you meant this?
```suggestion
          // -> [id{5*3} id{6}] -> split(0, 3)
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+          // well as the dimensions to the left of that.
+          continue;
+        }
+        break;
```
Why break here?

naoyam replied:
I think this part tries to find the domain next to the outer domain, excluding broadcasts. That domain must be the inner domain to keep propagating the contig domains.

zasdfgbnm replied:
oh yeah, never mind. I was thinking about something stupid.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+      // relative to outer
+      auto pos_after_outer = outer_pos + 1;
+      for (; pos_after_outer < ids.size(); pos_after_outer++) {
+        if (ids[pos_after_outer]->isBroadcast()) {
```
Should this be the following?
```suggestion
        if (ids[pos_after_outer]->isBroadcast() && pos_after_outer != inner_pos) {
```

naoyam replied:
I'm not sure if `ids[pos_after_outer]` can be a broadcast, but I don't think we would need `&& pos_after_output`. It's just trying to find a non-broadcast domain next to the outer domain, I think.

zasdfgbnm replied:
Is there any possibility that we have `bo, bi = split(b, 1)` here? If we are sure that we won't have such case, then yes, I think it is ok to not have the `&& ...` part.

csarofeen replied:
Yeah, I think this should be good to take in. Thanks.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        partial_projected_extent_[split->in()] =
+            getMaybePartialMappedExtent(split->inner());
+        // Clear to the left of split in since it partially maps
+        ids.erase(ids.begin(), ids.begin() + inner_pos);
```
I think inner_pos should already be 0 at this point? We can try the following:
```suggestion
        TORCH_INTERNAL_ASSERT(inner_pos == 0);
```

naoyam replied:
Yeah, that's my understanding.

csarofeen replied:
Agreed, but I'm officially scared of modifying this code because of a redundant line. Commented.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+      if (!hasPartialExtent(merge->out())) {
+        // No partial map in output, so simply map through to inputs.
+        ids[out_pos] = merge->outer();
+        ids.insert(ids.begin() + out_pos + 1, merge->inner());
```
nit:
```suggestion
        ids.insert(find_out_it + 1, merge->inner());
```

zasdfgbnm reviewed (commented):
Posting some of my reviews. Not done yet.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+        // and outer partial extent:
+        //   auto outer_partial_extent =
+        //   SimplifyingIrBuilder::divExpr(getMaybePartialMappedExtent(merge->out()),
+        //   merge->inner()->extent())
```
I don't think this can be completely symbolicly. Depending on the evaluated value of the `outer_partial_extent`, we might want to just remove the outer id (for example, if `outer_partial_extent` is 0 or 1), also, although `merge->out()` is guaranteed to be divisible by `merge->inner()->extent()`, I don't think `getMaybePartialMappedExtent(merge->out())` has the same good property. Depending on the runtime value whether it is or not divisible or not, we might need to make different decisions on whether to keep the outer/inner ID.

csarofeen replied:
I'm not following you. I think this could be done all symbolically but seems like not worth the effort right now. I agree we could want to keep or drop different dimensions, but we could track all of that in conditionals and evaluate them at runtime like anything else. You're right we may or may not want to remove that outer id depending on the sizes, but just throw if statements in and evaluate those. I'm missing why we can't do this.

If outer or inner is partial we just remove everything to the left, so even if we keep outer but it's partially mapped, I don't know I see the problem as long as we evaluate if it's partial or not at all mapped. We can only keep one of these dimensions on the left most side of the mapped dims. I believe all we'd be debating is if outer partially maps, or if inner partially maps.

csarofeen replied:
> I don't think getMaybePartialMappedExtent(merge->out()) has the same good property. Depending on the runtime value whether it is or not divisible or not, we might need to make different decisions on whether to keep the outer/inner ID.

I think with only processing divisible splits this would be the case (will push code for this). Effectively what we need to say for these passes is that we can only expand vectorization over dimensions with divisible splits. Which I think makes sense on our conversation earlier this week. So we are effectively always maintaining divisibility.

csarofeen replied:
Still thinking this through, interesting example:
```
            // [3*5, 7*11] split->(0, 5)
            // [3, 5, 7*11] merge->(2)
            // [3, 5*7*11] split(1, 5)
            // [3, 7*11, 5]
            // If we start by mapping the inner most dimension of the last
            // domain. we have:
            // [3, 7*11, 5] 5 full maps
            // [3, 5*7*11] 5 partial maps
            // Now we would try to propagate
            // [3, 5*7*11] -> [3, 5, 7*11]
            // Problem here is we would partial map 5 to 7*11 but it is not
            // divisible by 5, so we can't vectorize based on that value.
```

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+          partial_projected_extent_[merge->inner()] =
+              SimplifyingIrBuilder::minExpr(
+                  getMaybePartialMappedExtent(merge->out()),
+                  merge->inner()->extent());
```
I don't think this is correct. I feel that, for this case, we should just drop both the inner and the outer. For example, if you have (edited)
```
6 = merge(2, 3)
```
and
```
getMaybePartialMappedExtent(merge->out()) == 2
```
I don't think we can simply propagate this `2` to inner.

naoyam replied:
Why? Aren't we assuming the merge inner and outer are contiguous?

zasdfgbnm replied:
For example, if we have
```
T0[5, 2, 3]
T1[5, 6] = view(T0)
T2[5, 3, 2] = view(T1)
Tref[3, 5, 2] = transpose(T2)
```
For this case, can we vectorize T0 by 2? Note that the contiguity for T0 could be `(false, false, true)`

Edit:
Oh, I see what you mean by
> Aren't we assuming the merge inner and outer are contiguous?

If T0 is guaranteed to be contiguous, it is probably fine to vectorize T0 by 2. But having a partial extent not a factor of the extent makes me feel strange. I need to think more to see if I can find any counter examples.

zasdfgbnm replied:
I think I have a counterexample now. It is slightly modified from https://github.com/csarofeen/pytorch/pull/2124#discussion_r1005061667
```
T0[5, 2, 3]
T3[5, 2, 3] = set(T0)
T1[5, 6] = view(T3)
T2[5, 3, 2] = view(T1)
Tref[3, 5, 2] = transpose(T2)
```
So `T3` is guaranteed to be contiguous, but `T0` is not. If the contiguity of `T0` is `(false, false, true)`, then we get wrong result.

naoyam replied:
Yeah, I'm not sure how it would work in that case. The comment on `ContiguousInnerDimensionMapper` indicates the reference is expected to be contiguous, but not sure about other tensors.

csarofeen replied:
Within getExpandedVectorization (this also comes in as default_word_size)
```.cpp
  for (auto inp_or_out : vectorizable_inputs_outputs) {
    auto dtype_size = dataTypeSize(
        inp_or_out->dtype(), indexModeToDtype(runtime_info.getIndexMode()));

    max_expand_size = std::min(
        max_expand_size,
        SchedulerRuntimeInfo::max_alignment_size_in_byte / dtype_size);
    max_expand_size = std::min(
        max_expand_size, runtime_info.getMaxVectorizableWidth(inp_or_out));
    common_alignment_size = std::min(
        common_alignment_size, runtime_info.getAlignmentSize(inp_or_out));
  }
```

So if you have the reference Tref, and it partially maps to T0, T0's vectorization is capped anyways to because of the contiguity. This pass isn't a stand alone ground truth, it's limited to any tensor that maps to the inner most dimension's local vectorization size.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+            // Partially map inner dimension
+            ids[out_pos] = merge->inner();
+            partial_projected_extent_[merge->inner()] =
+                getMaybePartialMappedExtent(merge->out());
```
Similar to https://github.com/csarofeen/pytorch/pull/2124#discussion_r1004991114 and https://github.com/csarofeen/pytorch/pull/2124#discussion_r1004993740, I think we should check divisibility here.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+std::unordered_map<TensorView*, std::vector<IterDomain*>>::iterator
+ContiguousInnerDimensionsMapper::projectIdToRFactor(
+    TensorView* ref,
+    std::vector<IterDomain*> ids) {
```
I will not read this for now. I think all my reviews for `projectIdToRoot` applies to this symmetrically. Will come back and read this after these reviews are resolved.

naoyam replied:
Me too

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+  // Id's in consumer to clear from the mapped set due to broadcast
+  // concretization.
+  std::unordered_set<IterDomain*> consumer_ids_to_clear;
+  if (to->hasBroadcast()) {
```
I think this check is redundant here. We already have 
```
      if ((!c_id->isBroadcast() && !c_id->isTrivialReduction()) &&
          p_id->isBroadcast()) {
```
which I think is sufficient. And also, what if `to` doesn't have any broadcast but has some trivial reduction?

csarofeen replied:
Why do we care about trivial reduction? That shouldn't interfere so we should be able to ignore it. This loop is only tracking down concretized broadcasts, if it's not concretized we can ignore it.

Yes, `to->hasBroadcast()` can be removed, but I think it makes the code more clear that it's the condition we really care about for the following block.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+std::vector<ContiguousInnerDimensionsMapper> getAllVectorizedMapsOf(
+    TensorView* ref) {
+  std::vector<ContiguousInnerDimensionsMapper> mappers;
+  auto root_dom = ref->hasReduction() && ref->hasRFactor()
+      ? ref->getRootDomain()
+      : ref->getMaybeRFactorDomain();
```
I don't understand this part. If the reference tensor is a view output, we don't want to start from the rfactor domain?

csarofeen replied:
I took this and forgot about the comment from it. This is only in the has reduction case, not the view case. don't remember why, let me try to find the comment associated with it.

csarofeen replied:
https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/scheduler/utils.cpp#L1191-L1201

csarofeen replied:
Added the same comment here.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,207 @@
+  static ContiguousInnerDimensionsMapper map(
+      TensorView* reference,
+      std::vector<IterDomain*> ids,
+      std::shared_ptr<const ComputeAtMap> ca_map) {
```
Is ca_map used?

csarofeen replied:
Thanks, will double check and remove if not.

csarofeen replied:
Not used 😆

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+    auto root_i = of_tv_root_size - i - 1;
+    auto root_id = of_tv_root[root_i];
+
+    if (root_id->extent()->isOneInt() || root_id->isBroadcast()) {
+      if (projected_dims[projected_dims_i - 1]->sameAs(root_id)) {
+        --projected_dims_i;
+      }
+      continue;
+    }
```
So we are ignoring the contiguity of broadcast dimensions? Today, if we have a tensor of shape `(4, 5, 1, 6)` and stride `(5000, 1000, 1000, 1)`, we will have contiguity `(true, true, false, true)`, this will pass the check here, but I don't think we can vectorize 4.

csarofeen replied:
Contiguity should be bounded by `getExpandedVectorization` function as the input `default_word_size` or the loop on `vectorizable_inputs_outputs`. Limitation for each tensor based on their contiguity should be taken into consideration, just not as part of this analysis.

csarofeen replied:
Note to self to double check this

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,881 @@ namespace fuser {
+      break;
+    }
+  }
+  return vectorizable_dim_sizes;
```
Is there any reason why we return a vector of `Val*` instead of returning `SimplifyingIrBuilder::mulExpr` of these vals?

csarofeen replied:
No, I just feel strange about adding expressions generally to fusions during heuristics. Maybe I shouldn't.

csarofeen replied:
I no longer feel strange about doing that. I've started adding a slew of new expressions 😄

zasdfgbnm reviewed (commented):
posting more

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/compile_time_info.h:
```diff
@@ -86,6 +88,16 @@ class VectorizableInputsAndOutputs {
       CompileTimeEntryType::VECTORIZABLE_INPUTS_AND_OUTPUTS;
 };
 
+//! Entry type definition class for `VECTORIZE_MAPS`,
+//!  stores the vectorizable TensorViews on a fusion's inputs and outputs.
+class VectorizeMaps {
+ public:
+  using DataType =
+      std::vector<vectorize_helper::ContiguousInnerDimensionsMapper>;
+  static const CompileTimeEntryType EntryType =
+      CompileTimeEntryType::VECTORIZE_MAPS;
+};
```
According to https://github.com/csarofeen/pytorch/pull/2124#discussion_r1005043035, can we just cache a std::vector<Val*> for vector sizes? If so, could these cached sizes be used for vectorization validation?

csarofeen replied:
Vectorize maps are relative to a reference. We'd have to cache the reference too, and then if for some reason the vectorization was run an other the validation wouldn't work... yeah we might be able to do something but I think it'd be nicer to just wait until we have better indexing logic we can process.

csarofeen replied:
Note to self: vectorize maps being relative to a reference is actually a bit of a pain point in the end. However, I don't have a good solution. The issue I'm hitting right now is the reference has an inner most broadcast dimension, but that dimension is concretized in the fusion. Our logic actually doesn't detect that we should vectorize those dimensions but not the reference, instead it tries to vectorize a more inner dimension in the reference which obviously results in vectorize size of 1 (or it definitely should since otherwise it would be an issue).

csarofeen committed:
Update torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

Co-authored-by: Gao, Xiang <qasdfgtyuiop@gmail.com>

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
Merge branch 'vectorize' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
Apply suggestions from code review

Co-authored-by: Gao, Xiang <qasdfgtyuiop@gmail.com>

csarofeen committed:
Merge branch 'vectorize' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen commented:
Couple notes:
At some point we might want scheduling to be done on top of a model parallel schedule application, or a layout optimization pass.

For these types of situations we shouldn't just propagate between root and rfactor, but from domain.
When we do something like this, we also need to build out swizzle support so we don't try to vectorize through a swizzle dimension.
In this case we would also need to make sure whatever dimensions we're trying to merge from the reference to potentially expand the vectorization size is not a result of a non-divisible split.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
merge conflicts.

csarofeen committed:
Change vectorization analysis to be fractional and divisibility based.

csarofeen committed:
Don't propagate through non-divisible split, use ca map for extents.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

zasdfgbnm commented:
Is this ready for review again?

csarofeen commented:
Yes, updated the description with current status. Might be a bug somewhere, but overall approach and infrastructure is all there.

naoyam commented on torch/csrc/jit/codegen/cuda/test/test_gpu_utils.cpp:
```diff
@@ -238,6 +240,892 @@ TEST_F(NVFuserTest, FusionTVDomainGuard_CUDA) {
   TORCH_CHECK(tv->domain()->contiguity() == false_true);
 }
 
+namespace {
```
nit: These new tests are all about vectorization, so we should move them to their own test file.

csarofeen replied:
I'll do so next time I trigger a full rebuild ;-)

naoyam commented on torch/csrc/jit/codegen/cuda/ir_builder.cpp:
```diff
@@ -349,6 +349,27 @@ Val* SimplifyingIrBuilder::mulExpr(Val* lhs, Val* rhs) {
+  return IrBuilder::divExpr(lhs, rhs);
+}
+
+Val* SimplifyingIrBuilder::ceilDivExpr(Val* lhs, Val* rhs) {
```
Nice! Adding this has been a long-time todo item for me 😄

csarofeen committed:
Comments, minor fixes.

naoyam commented on torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:
```diff
@@ -1721,7 +1721,46 @@ TEST_F(NVFuserTest, FusionViewMagicSchedule5_CUDA) {
+  // TODO: Fix ref handling int pointwise scheduler to accept this program as
+  // one segment.
+  TORCH_CHECK(!executor_cache.getMostRecentKernelRuntime()->isSegmented());
+  TORCH_CHECK(executor_cache.getMostRecentExecutorInfo()
```
We should also verify that the vectorization factor is 2.

naoyam commented on torch/csrc/jit/codegen/cuda/test/test_gpu_view.cpp:
```diff
@@ -1721,7 +1721,46 @@ TEST_F(NVFuserTest, FusionViewMagicSchedule5_CUDA) {
+  executor_cache.profile(true);
+  auto cg_outputs = executor_cache.runFusionWithInputs({t0, t3});
+
+  // TODO: Fix ref handling int pointwise scheduler to accept this program as
```
Is this still a TODO item?

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  bool updated = true;
+  while (updated && i > 1) {
+    updated = false;
+    for (auto factor = 3; factor * factor < i; factor += 2) {
```
Shouldn't the bounds check be `factor * factor <= i`? For example, this doesn't seem to factorize at all when `i` is 9.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  auto simplified_set1 = set1;
+  auto simplified_set2 = set2;
+
+  simplified_set1.erase(1);
+  simplified_set2.erase(1);
```
Can the factor sets have `1`?

csarofeen replied:
At this point I don't remember, but removing them if they're there is never a problem. If I have some time will try to revisit and figure out if it's possible

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  std::multiset<int64_t> set1(vec1.begin(), vec1.end());
+  std::multiset<int64_t> set2(vec2.begin(), vec2.end());
+
+  set1.erase(1);
+  set2.erase(1);
```
Same here

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,487 @@
+      if (numerator_ == nullptr) {
+        numerator_ = numerator_val;
+      } else {
+        numerator_ = IrBuilder::mulExpr(numerator_, numerator_val);
```
Can we just use `SimplifyingIrBuilder`?

csarofeen replied:
Yeah, I'm not sure why we wouldn't just always use simplifying ir builder, except that it might be harder when we read the code to know where the constant value comes from. We probably should add a disable option to disable simplifying IR expression.

csarofeen replied:
Went through and tried to update all uses to IrBuilder that made sense to move to simplifyingIrBuilder

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  auto simplified_set2 = set2;
+
+  for (auto fact1 : set1) {
+    auto fact2_it = simplified_set2.find(fact1);
```
Not sure what the context of this function would be, but do we need to do the comparison possibly with actual values rather than just comparing pointers?

csarofeen replied:
WDYM?

naoyam replied:
Actually don't remember why I asked this question, but I think I was wondering about the definition of sameness, so for example we could also use ExpressionEvaluator to find "same" values, but I assume just using object addresses is fine for this use case.

csarofeen replied:
Object addresses are used in this one. There's the other simplifier which uses explicit const values. This approach is processing compile time constant values separately from the symbolic, so the assumption is if this is used, it's only possible to do by `Val*` not by Expression Evaluation.

naoyam reviewed (commented):
Just some minor comments for now. Will add more comments.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+        if (ids[pos_after_outer]->isBroadcast() &&
+            pos_after_outer != inner_pos) {
+          // Skip broadcast axes as they must not have been concretized in the
+          // reference. We remove dimensions that underwent a concretization as
```
Where is the removal done?

naoyam replied:
Ok, found it in `propagateC2P`. Would be great if mentioned in the comment.

csarofeen replied:
Yeah, will add, this turns out to be a pain point with how we use the reference of schedulers... https://github.com/csarofeen/pytorch/pull/2124#discussion_r1024444020

Seems like we might have to track sizes through resolved broadcasts but not unresolved broadcasts, and not remove the mappings when going back from resolved to unresolved, but understand a mapping might be on a broadcast dim but isn't a valid vectorizable dim for that tensor.

Just more notes to myself, this is pretty painful and might require some changes across scheduling and vectorization detection.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+      }
+
+      auto& inner_mapping = getMappedExtent(split->inner());
+      if (inner_mapping.isZero()) {
```
Is the outer extent also checked somewhere?

csarofeen replied:
No, I'm not actually really sure how much value this provides. Theoretically prevents us from processing zeroes, but in practice might not be worth the few lines of code it is.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+// that can be used. Otherwise return the concrete ID extent which makes
+// simplification pass easier as the same factor will be used more consistently
+// where possible.
+Val* commonOrConstExtent(
```
Maybe this could be just done when building the CA map for all disjoint sets. It seems like generally useful information.

csarofeen replied:
Yes, agreed, I just didn't want to try and refactor the whole codebase right now to do so.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  for (auto* expr : transform_exprs) {
+    if (Split* split = dynamic_cast<Split*>(expr)) {
+      // Initialize state
+      auto find_outer_it = ids.begin();
+      auto outer_pos = ids.size();
+      auto find_inner_it = ids.begin();
+      auto inner_pos = ids.size();
+
+      // Removes all entries to the left of provided `it`, if `it` is not
+      // ids.begin(). Updates all state of finding outer and inner in the ids
+      // vector after erasing.
+      auto clear_left_of = [&find_outer_it,
```
Not a blocker, but I'm strongly concerned about this code. These iterators are invalidated when we do something like `ids.erase(...)`, and a common error I've seen is to keep using such invalid iterators. Here, it is valid as long as the modification of `ids` is done by `clear_left_of`, but we also do `ids.erase(...)`, and we must not use the iterators after that. While I don't see any error in this code, it'd really easy to accidentally use them. Perhaps, we should wrap `ids` in a small struct and provide accessors with proper error checks.

I could clean it up later if you'd like. Feel free to leave it as is.

csarofeen replied:
Yeah, I definitely agree, I frequently get in trouble managing my iterators manually which is why I implemented this. An overloaded erase could easily call this function with `clear_left_of(ids.begin());` so wrapping ids makes a lot of sense.

csarofeen replied:
Going to leave for now, want to see if I can get this cleaned up enough to get merged in.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,487 @@
+      return projected_extent_.at(id);
+    }
+    ProjectedExtent init_extent;
+    init_extent.multiplyDenominatorValue(id->fusion()->oneVal());
```
nit: Should this be done in a constructor of `ProjectedExtent`?

csarofeen replied:
Isn't necessary to initialize the Projected extent anymore.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+
+      auto out_mapping = getMappedExtent(merge->out());
+      if (out_mapping.isZero()) {
+        ids.erase(ids.begin(), ids.begin() + out_pos + 1);
```
Shouldn't this be `ids.erase(ids.begin(), ids.begin() + out_pos + 2)`? The inner ID was inserted just before this part of the code, and if the out is zero, both the outer and inner IDs should be discarded, not?

csarofeen replied:
Moved the insertion to after this block, thanks.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+      const auto out_numerator = out_mapping.getNumerator();
+      const auto out_denominator = out_mapping.getDenominator();
+
+      auto out_bigger_than_inner = IrBuilder::geExpr(
```
nit: `gtExpr`?

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+          FusionGuard::getCurFusion()->oneVal()));
+      // otherwise use the out mapping extent
+      inner_mapping.maybeMul(
+          SimplifyingIrBuilder::notExpr(out_bigger_than_inner), out_mapping);
```
Why not just putting `out_mapping` instead of `oneVal` in the above `where` expr?

csarofeen replied:
out_mapping isn't a Val*, it's a Projected Extent. So this preserves the numerator/denominator in a way that's easier to simplify rather than having a complex Val* expression.

naoyam reviewed (commented):
Some more comments

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+        // relative to outer
+        auto pos_after_outer = outer_pos + 1;
+        for (; pos_after_outer < ids.size(); pos_after_outer++) {
+          if (ids[pos_after_outer]->isBroadcast()) {
```
It doesn't seem necessary but `projectToRoot` has `pos_after_outer != inner_pos`

csarofeen replied:
Thanks, I think it might be necessary, as it's fine to have an inner position as broadcast, if that's not the inner position than we can just ignore the broadcast, but if it is the inner position we can pick it up.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+// logic of split and merge as the reverse mapping of merge looks a lot like
+// split and vice versa.
+std::unordered_map<TensorView*, std::vector<IterDomain*>>::iterator
+ContiguousInnerDimensionsMapper::projectIdToRFactor(
```
This is almost exactly the same as `projectIdToRoot`. Not a blocker, but we should consider refactor the two versions.

csarofeen replied:
Agreed we might be able to make it cleaner.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  return projected_rfactor_ids_.emplace(ref, ids).first;
+}
+
+void ContiguousInnerDimensionsMapper::propagateC2P(
```
When this is `auto to = transpose(from)`, don't we need to do something different? Looks like all the contig IDs of a consumer would be propagated to its producer even if the expr is a transpose.

naoyam replied:
I found there's something related in `getContigVectorSizesOf`. It seems that transpose is not taken into consideration at the time when `ContiguousInnerDimensionsMapper` is built, but schedulers should remove transposed dimensions out of vectorization candidates. Is it correct?

csarofeen replied:
No, they shouldn't, as we can have reversible transposes:
```
t0[i0, i1]
t1[i1, i0] = transpose(t0)
t2[i0, i1] = transpose(t1)
```

So as long as we have matches, we just use the references ordering. We check for consistency of ordering in `getContigVectorSizeOf`:
```
    // Mapping order isn't correct, cannot expand vectorization dimension.
    if (!projected_dims[--projected_dims_i]->sameAs(root_id)) {
      break;
    }
```

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+    return {};
+  }
+
+  // Filter out mismatched contiguity info
```
When this happens?

csarofeen replied:
Don't think it should, will change to an assert.

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+  if (dim_info.empty()) {
+    return 1;
+  }
+
```
`// Reverse the size vector to traverse from the innermost dimensions`

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1092 @@ namespace fuser {
+    TORCH_INTERNAL_ASSERT(
+        numerator_optional->isInt() && denominator_optional->isInt() &&
+            extent_optional->isInt(),
+        "Vectorization heuristic expentin integer values.");
```
typo: expentin -> expecting?

naoyam commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -191,97 +1113,27 @@ size_t expandVectorizationToContigMergedDomains(
-
-  return word_size;
+  max_supported_vector_size =
+      std::min(max_supported_vector_size, max_expand_size);
```
nit: this should not be needed.

naoyam reviewed (approved):
Finished reading the code. Approved the PR, though I'm not very confident. My comments are mostly suggestions about clarity and code cleanup, none of which is a blocker.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
Merge conflict fix.

csarofeen committed:
Comments.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp:
```diff
@@ -17,171 +18,1096 @@ namespace fuser {
+  while (updated && i > 1) {
+    updated = false;
+    for (auto factor = 3; factor * factor <= i; factor += 2) {
+      if (i % factor == 0) {
```
should this `if` be a `while`?

zasdfgbnm replied:
For example, if I have `i = 3*3*3*5*5*5`, this algorithm will find 9, 15 as a factor, but don't we want to have just 3 and 5?

csarofeen replied:
Thanks, missed a break 😳

csarofeen replied:
Yeah, that was not written well... thanks for flagging it.

csarofeen committed:
Merge branch 'devel' of https://www.github.com/csarofeen/pytorch into vectorize

csarofeen committed:
Fix brute force factorization.

csarofeen committed:
Vectorize - step 2 max spanning tree fix (#2196)

- Extract out extent propagation.
- Vectorization pass implements its own spanning tree.
- If any vectorized tensor doesn't have a mapped dimension return size 1 vectorize.

csarofeen committed:
Lint.

zasdfgbnm commented on torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.h:
```diff
@@ -1,37 +1,591 @@
+      return mapped_root_ids_.size() <
+          other_mapped_domain.mapped_root_ids_.size() &&
+          mapped_rfactor_ids_.size() <
+          other_mapped_domain.mapped_rfactor_ids_.size();
```
I think this should be:
```C++
if (mapped_root_ids_.size() != other_mapped_domain.mapped_root_ids_.size()) {
  return mapped_root_ids_.size() < other_mapped_domain.mapped_root_ids_.size()
}
return mapped_rfactor_ids_.size() < other_mapped_domain.mapped_rfactor_ids_.size();
```

csarofeen replied:
If they didn't agree I think that would be a fundamental problem. I'd rather throw in:
```
TORCH_INTERNAL_ASSERT(mapped_root_ids_.size() < other_mapped_domain.mapped_root_ids_.size() == 
mapped_rfactor_ids_.size() < other_mapped_domain.mapped_rfactor_ids_.size())
```

zasdfgbnm replied:
Oh nvm, I thought when a tensor has no rfactor domain, then `mapped_rfactor_ids_` would be empty

csarofeen replied:
Ahh! that's understandable. It should really be `maybe_mapped_rfactor_ids_` lol.

csarofeen replied:
@zasdfgbnm took one more stab at this. Please check the last commit here when you get a chance.

csarofeen replied:
Flipping `<` with `>` still doesn't fail though 😆

csarofeen committed:
Max spanning tree of vectorization again.
